orttraining/orttraining/core/graph/pipeline_transformer.cc

```diff
@@ -691,33 +722,359 @@ Status TransformGraphForPipeline(
+    ONNX_NAMESPACE::AttributeProto tag;
+    tag.set_name("tag");
+    tag.set_type(ONNX_NAMESPACE::AttributeProto_AttributeType::AttributeProto_AttributeType_INT);
+    // currently hard-coded all tag to be 0. May need to change when multiple GPU stream is used.
+    tag.set_i(static_cast<int64_t>(0));
+
+    ONNX_NAMESPACE::AttributeProto element_types;
+    element_types.set_name("element_types");
+    element_types.set_type(ONNX_NAMESPACE::AttributeProto_AttributeType::AttributeProto_AttributeType_INTS);
+
+    // for each edge in this group, perform edge cut
+    for (auto& id : edgeIds) {
+      // find node whose output contains id.node_arg_name
+      auto producer_node = graph.GetMutableProducerNode(id.node_arg_name);
+      if (producer_node == nullptr) {
+        ORT_THROW("Cannot find producer node of node_arg with name: ", id.node_arg_name, ". Wrong cutting infomation.");
+      }
+
+      // once we find out the producer node for id.node_arg_name, find which output index that leads
+      // to id.node_arg_name
+      int upstream_nodes_output_index{-1};
+      producer_node->ForEachWithIndex(
+          producer_node->OutputDefs(),
+          [&](const NodeArg& def, size_t index) {
+            if (def.Name() == id.node_arg_name) {
+              upstream_nodes_output_index = index;
+            }
+            return Status::OK();
+          });
+
+      if (upstream_nodes_output_index < 0) {
+        ORT_THROW("Node with name: ", producer_node->Name(),
+                  " doesn't have an output node_arg with name ", id.node_arg_name);
+      }
+
+      size_t idx = static_cast<size_t>(upstream_nodes_output_index);
+
+      // original node_arg pointer from the origin graph. This serves as the key in the
+      // updated_node_arg map and any reference for original node_arg name
+      auto original_node_arg = producer_node->MutableOutputDefs()[idx];
+
+      // updated node_arg pointer from previous partition. This is the new arg_node the
+      // current inserted send node will take as input node_arg.
+      auto updated_node_arg = producer_node->MutableOutputDefs()[idx];
+      auto exiting_updated_node_arg = updated_node_args.find(original_node_arg);
+      if (exiting_updated_node_arg != updated_node_args.end()) {
+        updated_node_arg = exiting_updated_node_arg->second;
+      }
+
+      send_input_args.push_back(updated_node_arg);
+
+      auto dtype = original_node_arg->TypeAsProto()->tensor_type().elem_type();
+      switch (dtype) {
+        case ONNX_NAMESPACE::TensorProto_DataType_FLOAT:
+          element_types.add_ints(static_cast<int64_t>(1));
+          break;
+        default:
+          // Assume all tensors are of type float.
+          // TODO: update if graph supports other data type.
+          ORT_THROW("pipeline partition unsupported 'type' value: ", dtype);
+      }
+
+      auto& new_receive_output = CreateNodeArg(graph, updated_node_arg);
+      auto old_shape = *(updated_node_arg->Shape());
```

Comment:
>auto [](start = 6, length = 4)

const auto& ?

Reply:
I am not sure if it is legit that two nodes share the same shape object -- updated_node_arg->Shape() returns its shape object, and new_receive_output.SetShape(old_shape) function takes old_shape obj as a reference instead of a copy. So if old_shape is a reference to *updated_node_arg->Shape(), it means both updated_node_arg and new_receive_output shares the same shape obj. 

I tested in the bert application with const auto&, it seems to work. But I am not sure whether if it is because I got lucky. 

So just want to confirm before I make the change.

---
In reply to: [428259076](https://github.com/microsoft/onnxruntime/pull/3996#discussion_r428259076) [](ancestors = 428259076)